### PR TITLE
Add menu option and hotkeys to move controllers/effects

### LINKED ISSUE #5773: Controller Rack - Ability to move controller modules up and down
Add the ability to move controller modules in the Controller Rack up and down.

### Benefits....
Allows you to organize / order them based on your setup preferences
Can allow you to separate LFO Controllers from Peak Controller modules

### Solutions.....
1.) Create two context menu items "Move Up" and "Move Down"

2.) Add ability to move controller modules by dragging with mouse 
(using similar code already used in SE & BBE that allows tracks to moved by mouse)
Wish this approach of moving items was used more throughout LMMS just makes workflow so much faster / easier.
I feel like I would move things so much more if it was, not for the benefit of my songs, but just because I can :)
But seriously though, can get a little tedious having to keep on opening context menus over and over if needing to move items great distances.

3.) Best of both worlds, adding both methods (I am sure I will get an argument or two for this statement) :)

Anyways, I could have sworn this already existed in the context menu?
May be just thinking of effect modules (which can also benefit from mouse drag moving).

### PR DESCRIPTION
**Pull request by Rossmaxx**
Supersedes #6154 and closes #5773.

from the original PR body.

> This commit adds UI changes, signals, and handlers required for moving controllers up and down in a controller rack. The feature was requested in the issue https://github.com/LMMS/lmms/issues/5773. The PR also adds hotkeys for moving controllers and effects in their respective racks, with a combination of ALT + Up/Down.

> The implementation is similar to the one used with EffectRacks. Quite small change, good for a first issue to solve (as it was mentioned in the issue itself)

> Added move up and move down buttons to context menu
Added signals that'll message the requested move
Added functions in the ControllerRackView that'll handle those signals
Added QActions for the hotkeys to move Controllers & Effects

### CODE REVIEW

**Comment by Rossmaxx** [comment]
for some reason, the controller rack's moveUp and moveDown functions doesn't seem to work, though i get the right click menu.

**Comment by Rossmaxx** [comment]
For some reason, the signals are not working. Here's some debug logs i get. 
```bash
QObject::connect: No such signal lmms::gui::ControllerView::moveUp(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:172
QObject::connect: No such signal lmms::gui::ControllerView::moveDown(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:174
QObject::connect: No such signal lmms::gui::ControllerView::moveUp(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:172
QObject::connect: No such signal lmms::gui::ControllerView::moveDown(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:174
QObject::connect: No such signal lmms::gui::ControllerView::moveUp(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:172
QObject::connect: No such signal lmms::gui::ControllerView::moveDown(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:174
QObject::connect: No such signal lmms::gui::ControllerView::moveUp(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:172
QObject::connect: No such signal lmms::gui::ControllerView::moveDown(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:174
QObject::connect: No such signal lmms::gui::ControllerView::moveUp(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:172
QObject::connect: No such signal lmms::gui::ControllerView::moveDown(ControllerView* ) in D:\BuildAndGit\lmms\src\gui\ControllerRackView.cpp:174
```

**Comment by zonkmachine** [comment]
Tested. Context menu and hotkeys works fine. :+1:

**Comment by zonkmachine** [comment]
Is this pr based on https://github.com/LMMS/lmms/pull/6154 ?

**Comment by zonkmachine** [comment]
> I asked the original author and took over the PR because they seem to have lost interest.

I don't see his name in the commit logs. I may have missed that but you should credit the original author in the commit history. I don't know what is the best way to do it but sometimes its just easier to push a bunch of changes to the original PR. Or you can keep the original commit with the original author intact and push it to a new branch with your added stuff to follow. Or you make sure to add a message in the final commit message "Based on original work by: ...".

**Comment by Rossmaxx** [comment]
I actually copy pasted the diff so you won't see his name. So what's the best way to credit him then?

**Comment by zonkmachine** [comment]
> I actually copy pasted the diff so you won't see his name. So what's the best way to credit him then?

I wonder too so I bumped #dev-support.

**Comment by Rossmaxx** [comment]
I did something to try credit @ejaaskel per request from zonk. For some reason, looks like @michaelgregorius is getting some credit too due to a mess up i did with git.

File: src/gui/ControllerRackView.cpp [hunk 1]
```diff
@@ -132,6 +133,33 @@ void ControllerRackView::deleteController( ControllerView * _view )
 	song->removeController( c );
 }
 
+void ControllerRackView::moveUp(ControllerView* view)
+{
+	// cannot delete the first module
```
Comment (michaelgregorius) [inline src/gui/ControllerRackView.cpp:138]: It seems like this comment does not belong here in the context of moving things. Guess it should be somthing like "Cannot move the first module further up" or similar.

**Comment by michaelgregorius** [comment]
@Rossmaxx, the best way to keep everything intact would be something along the following lines:
1. Add https://github.com/ejaaskel/lmms.git as a remote to your (local) repository.
2. Checkout the branch [movable-controllers](https://github.com/ejaaskel/lmms/tree/movable-controllers) by @ejaaskel.
3. Optional: Merge the current master into the branch if necessary.
4. Add your changes on top.
5. Push the branch to your remote at GitHub.
6. Create a new pull request from your remote.

This way GitHub should see all the involved commits and credit @ejaaskel appropriately.

**Comment by zonkmachine** [comment]
> Optional: Merge the current master into the branch if necessary.

And squash it down to one commit probably.

**Comment by michaelgregorius** [comment]
> > Optional: Merge the current master into the branch if necessary.
> 
> And squash it down to one commit probably.

This can also be done during the final merge by using the option "Squash and merge" in GitHub. At least that's what I am using after being told to do so. :sweat_smile:

**Comment by Rossmaxx** [comment]
Noted

**Comment by Rossmaxx** [comment]
@michaelgregorius i fixed it using an interactive rebase.

File: src/gui/ControllerRackView.cpp (hunk 1 above)
Comment (Rossmaxx) [reply src/gui/ControllerRackView.cpp]: fixed with https://github.com/LMMS/lmms/pull/7139/commits/3d49888b825a2b0d1986448afabd6e6c4b3d4755

**Comment by Rossmaxx** [comment]
@michaelgregorius can i get an approval here?

File: src/gui/ControllerRackView.cpp [hunk 2]
```diff
@@ -141,9 +169,26 @@ void ControllerRackView::onControllerAdded( Controller * controller )
...
+	connect(controllerView, SIGNAL(moveDown(lmms::gui::ControllerView* )), this, 
+		SLOT(moveDown(lmms::gui::ControllerView* )), Qt::QueuedConnection);
 	connect( controllerView, SIGNAL(deleteController(lmms::gui::ControllerView*)),
 		 this, SLOT(deleteController(lmms::gui::ControllerView*)), Qt::QueuedConnection );
 
+	QAction* moveUpAction = new QAction(controllerView);
+	moveUpAction->setShortcut(Qt::Key_Up | Qt::AltModifier);
+	moveUpAction->setShortcutContext(Qt::WidgetShortcut);
+	connect(moveUpAction, SIGNAL(triggered()), controllerView, SLOT(moveUp()));
+	controllerView->addAction( moveUpAction );
```
Comment (michaelgregorius) [inline src/gui/ControllerRackView.cpp:183]: Unnecessary white space in new code. :wink:

**Comment by michaelgregorius** [comment]
@Rossmaxx, with regards to the functionality everything is working for me. However, it would be nice if there was an indication of what element is currently selected so that the users know which element will be affected by the action.

I think the code might also get simpler if there was a concept of a selected element because some of the signals would not be needed in that case. The shortcuts would be defined for the widget that presents the list and if activated the widget would search for the selected element and work on that. With the current code a list element knows how to move itself (by emitting a signal which is caught by the parent) which is a little bit unintuitive IMO.

However, I guess that's not really in the scope of this PR.

**Review by michaelgregorius — APPROVED** [review]

**Comment by sakertooth** [comment]
Sorry, forgot about this. I saw a couple of problems with this that I wanted to mention. Give me a second.

File: src/gui/ControllerView.cpp [hunk 3]
```diff
@@ -173,7 +177,11 @@ void ControllerView::modelChanged()
 
 void ControllerView::contextMenuEvent( QContextMenuEvent * )
 {
-	QPointer<CaptionMenu> contextMenu = new CaptionMenu( model()->displayName(), this );
+	Controller* c = castModel<Controller>();
+	QPointer<CaptionMenu> contextMenu = new CaptionMenu(c->name(), this);
+	contextMenu->addAction(embed::getIconPixmap("arp_up"), tr("Move &up"), this, SLOT(moveUp()));
```
Comment (sakertooth) [inline src/gui/ControllerView.cpp:182]:
```suggestion
	contextMenu->addAction(embed::getIconPixmap("arp_up"), tr("Move &up"), this, &ControllerView::moveUp);
```

File: src/gui/ControllerView.cpp [hunk 4]
```diff
@@ -173,7 +177,11 @@ void ControllerView::modelChanged()
...
+	contextMenu->addAction(embed::getIconPixmap("arp_down"), tr("Move &down"), this, SLOT(moveDown()));
```
Comment (sakertooth) [inline src/gui/ControllerView.cpp:183]:
```suggestion
	contextMenu->addAction(embed::getIconPixmap("arp_down"), tr("Move &down"), this, &ControllerView::moveDown);
```

File: src/gui/EffectRackView.cpp [hunk 5]
```diff
@@ -177,6 +178,19 @@ void EffectRackView::update()
 			connect( view, SIGNAL(deletePlugin(lmms::gui::EffectView*)),
 				this, SLOT(deletePlugin(lmms::gui::EffectView*)),
 							Qt::QueuedConnection );
+
+			QAction* moveUpAction = new QAction(view);
+			moveUpAction->setShortcut(Qt::Key_Up | Qt::AltModifier);
+			moveUpAction->setShortcutContext(Qt::WidgetShortcut);
+			connect(moveUpAction, SIGNAL(triggered()), view, SLOT(moveUp()));
```
Comment (sakertooth) [inline src/gui/EffectRackView.cpp:185]:
```suggestion
			connect(moveUpAction, &QAction::triggered, view, &EffectView::moveUp);
```

File: src/gui/EffectRackView.cpp [hunk 6]
```diff
@@ -177,6 +178,19 @@ void EffectRackView::update()
...
+			view->addAction(moveUpAction);
+
+			QAction* moveDownAction = new QAction(view);
+			moveDownAction->setShortcut(Qt::Key_Down | Qt::AltModifier);
+			moveDownAction->setShortcutContext(Qt::WidgetShortcut);
+			connect(moveDownAction, SIGNAL(triggered()), view, SLOT(moveDown()));
```
Comment (sakertooth) [inline src/gui/EffectRackView.cpp:191]:
```suggestion
			connect(moveDownAction, &QAction::triggered, view, &EffectView::moveDown);
```

File: src/gui/ControllerRackView.cpp [hunk 7]
```diff
@@ -132,6 +133,33 @@ void ControllerRackView::deleteController( ControllerView * _view )
...
+	for (auto it = m_controllerViews.begin(); it != m_controllerViews.end(); ++it, ++i)
+	{
+		if (*it == view) { break; }
+	}
+	ControllerView* temp = m_controllerViews[i - 1];
+	m_controllerViews[i - 1] = view;
+	m_controllerViews[i] = temp;
+
+	m_scrollAreaLayout->removeWidget(view);
+	m_scrollAreaLayout->insertWidget(i - 1, view);
```
Comment (sakertooth) [inline src/gui/ControllerRackView.cpp:151]:
```suggestion
	if (view == m_controllerViews.first()) { return; }
	
	const auto index = std::distance(m_controllerViews.begin(), view);
	std::swap(m_controllerViews[index - 1], m_controllerViews[index]);
	m_scrollAreaLayout->removeWidget(view);
	m_scrollAreaLayout->insertWidget(index - 1, view);
```

File: src/gui/ControllerRackView.cpp (hunk 2 above)
Comment (Rossmaxx) [reply src/gui/ControllerRackView.cpp]: removed the extra space.

**Comment by sakertooth** [comment]
Had a feeling I would break build..., sorry, just one second.

**Comment by sakertooth** [comment]
I'm going to push my fixes directly here since its easier (and safer) compared to reviewing it on Github again.

**Review by sakertooth — APPROVED** [review]

**Pull request merged**